src/adjustproxyconnection.cpp

```diff
@@ -0,0 +1,196 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "adjustproxyconnection.h"
+#include "leakdetector.h"
+#include "logger.h"
+#include "mozillavpn.h"
+#include "qmlengineholder.h"
+#include "networkrequest.h"
+
+#include <QFile>
+#include <QHostAddress>
+#include <QTcpSocket>
+#include <QUrl>
+#include <QUrlQuery>
+
+constexpr const char* HTTP_RESPONSE =
+    "HTTP/1.1 <status>\nContent-Type: "
+    "application/json\n\n<message>\n";
+
+namespace {
+Logger logger(LOG_ADJUST, "AdjustProxyConnection");
+}  // namespace
+
+AdjustProxyConnection::AdjustProxyConnection(QObject* parent,
+                                             QTcpSocket* connection)
+    : QObject(parent), m_connection(connection) {
+  MVPN_COUNT_CTOR(AdjustProxyConnection);
+
+  logger.debug() << "New connection received";
+
+  Q_ASSERT(m_connection);
+  connect(m_connection, &QTcpSocket::readyRead, this,
+          &AdjustProxyConnection::readData);
+}
+
+AdjustProxyConnection::~AdjustProxyConnection() {
+  MVPN_COUNT_DTOR(AdjustProxyConnection);
+  logger.debug() << "Connection released";
+}
+
+void AdjustProxyConnection::readData() {
+  Q_ASSERT(m_connection);
+  QByteArray input = m_connection->readAll();
+  m_buffer.append(input);
+
+  switch (m_state) {
+    case ProcessingState::NotStarted:
+      if (m_buffer.isEmpty()) {
+        return;
+      }
+      processFirstLine();
+      [[fallthrough]];
+    case ProcessingState::FirstLineDone:
+      if (m_buffer.isEmpty()) {
+        return;
+      }
+      processHeaders();
+      [[fallthrough]];
+    case ProcessingState::HeadersDone:
+      processParameters();
+      [[fallthrough]];
+    case ProcessingState::ParametersDone:
+      filterParametersAndForwardRequest();
+      break;
+    default:
+      Q_ASSERT(false);
+      logger.debug() << "Unknown ProcessingState: " << m_state;
+  }
+}
+
+void AdjustProxyConnection::processFirstLine() {
+  int pos = m_buffer.indexOf("\n");
+  if (pos == -1) {
+    return;
+  }
+
+  QByteArray line = m_buffer.left(pos);
+  m_buffer.remove(0, pos + 1);
+
+  QList<QByteArray> parts = line.split(' ');
+  if (parts.length() < 2) {
+    logger.info() << "Invalid HTTP request";
+    m_connection->close();
+    return;
+  }
+
+  m_method.append(parts[0]);
+  m_route = parts[1].trimmed();
+
+  m_state = ProcessingState::FirstLineDone;
+}
+
+void AdjustProxyConnection::processHeaders() {
+  while (true) {
+    int pos = m_buffer.indexOf("\n");
+    if (pos == -1) {
+      break;
```

Comment:
This should be a return. We don't want to set status = `HeadersDone` if we have an unfinished line.